Invalid move in S-chess game

Case described in https://github.com/gbtami/pychess-variants/issues/1746

If I add this to test.py test_legal_moves() unittest

```
        # Seirawan gating
        fen = "rnbq3r/pp2bkpp/8/2p1p2K/2p1P3/8/PPPP1PPP/RNB4R[EHeh] b ABCHabcdh - 0 10"
        result = sf.legal_moves("seirawan", fen, [])
        self.assertIn("c8g4h", result)

```
the result is
```
======================================================================
FAIL: test_legal_moves (__main__.TestPyffish.test_legal_moves)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/tamas/Fairy-Stockfish/test.py", line 393, in test_legal_moves
    self.assertIn("c8g4h", result)
AssertionError: 'c8g4h' not found in ['c4c3', 'a7a6', 'b7b6', 'g7g6', 'h7h6', 'a7a5', 'b7b5', 'g7g5', 'b8a6', 'b8c6', 'b8d7', 'e7h4', 'e7g5', 'e7d6', 'e7f6', 'e7f8', 'c8h3', 'c8g4', 'c8f5', 'c8e6', 'c8d7', 'h8e8', 'h8f8', 'h8g8', 'd8d2', 'd8d3', 'd8d4', 'd8a5', 'd8d5', 'd8b6', 'd8d6', 'd8c7', 'd8d7', 'd8e8', 'd8f8', 'd8g8', 'f7e6', 'f7f6', 'f7g8', 'f7e8', 'f7f8']

----------------------------------------------------------------------
Ran 21 tests in 1.201s

FAILED (failures=1)

```

--- a/src/position.cpp
+++ b/src/position.cpp
@@ -384,17 +384,16 @@ Position& Position::set(const Variant* v, const string& fenStr, bool isChess960,
                                          : make_square(castling_king_file(), castling_rank(c));
               // Skip invalid castling rights
               if (!(castlingKings & st->castlingKingSquare[c]))
-              {
                   st->castlingKingSquare[c] = SQ_NONE;
-                  continue;
-              }
           }
 
           // Set gates (and skip castling rights)
           if (gating())
           {
-              st->gatesBB[c] |= rsq;
-              if (token == 'K' || token == 'Q')
+              // Only add gates for occupied squares
+              if (pieces(c) & rsq)
+                  st->gatesBB[c] |= rsq;
+              if ((token == 'K' || token == 'Q') && st->castlingKingSquare[c] != SQ_NONE)
                   st->gatesBB[c] |= st->castlingKingSquare[c];
               // Do not set castling rights for gates unless there are no pieces in hand,
               // which means that the file is referring to a chess960 castling right.
@@ -402,7 +401,10 @@ Position& Position::set(const Variant* v, const string& fenStr, bool isChess960,
                   continue;
           }
 
-          if (castling_enabled() && (castling_rook_pieces(c) & type_of(piece_on(rsq))) && color_of(piece_on(rsq)) == c)
+          // Only add castling right if both king and rook are on expected squares
+          if (   castling_enabled()
+              && st->castlingKingSquare[c] != SQ_NONE
+              && (castling_rook_pieces(c) & type_of(piece_on(rsq))) && color_of(piece_on(rsq)) == c)
               set_castling_right(c, rsq);
       }
 

--- a/test.py
+++ b/test.py
@@ -337,6 +337,11 @@ class TestPyffish(unittest.TestCase):
         result = sf.legal_moves("shogun", SHOGUN, ["c2c4", "b8c6", "b2b4", "b7b5", "c4b5", "c6b8"])
         self.assertIn("b5b6+", result)
 
+        # Seirawan gating but no castling
+        fen = "rnbq3r/pp2bkpp/8/2p1p2K/2p1P3/8/PPPP1PPP/RNB4R[EHeh] b ABCHabcdh - 0 10"
+        result = sf.legal_moves("seirawan", fen, [])
+        self.assertIn("c8g4h", result)
+
         # In Cannon Shogi the FGC and FSC can also move one square diagonally and, besides,
         # move or capture two squares diagonally, by leaping an adjacent piece. 
         fen = "lnsg1gsnl/1rc1kuab1/p1+A1p1p1p/3P5/6i2/6P2/P1P1P3P/1B1U1ICR1/LNSGKGSNL[] w - - 1 3"
@@ -458,6 +463,12 @@ class TestPyffish(unittest.TestCase):
         fen0 = "reb1k2r/ppppqppp/2nb1n2/4p3/4P3/N1P2N2/PB1PQPPP/RE2KBHR[h] b KQkqac - 2 6"
         fen1 = "reb2rk1/ppppqppp/2nb1n2/4p3/4P3/N1P2N2/PB1PQPPP/RE2KBHR[h] w KQac - 3 7"
         result = sf.get_fen("seirawan", fen0, ["e8g8"])
+        self.assertEqual(result, fen1)
+
+        # handle invalid castling/gating flags
+        fen0 = "rnbq3r/pp2bkpp/8/2p1p2K/2p1P3/8/PPPP1PPP/RNB4R[EHeh] b QBCEHabcdk - 0 10"
+        fen1 = "rnbq3r/pp2bkpp/8/2p1p2K/2p1P3/8/PPPP1PPP/RNB4R[EHeh] b ABCHabcdh - 0 10"
+        result = sf.get_fen("seirawan", fen0, [])
         self.assertEqual(result, fen1)
 
         result = sf.get_fen("chess", CHESS, [], True, False, False)

